pkg/filters/oidcadaptor/oidcadaptor.go

```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package oidcadaptor
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/MicahParks/keyfunc"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/google/uuid"
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/filters"
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/protocols/httpprot"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+//https://openid.net/specs/openid-connect-core-1_0.html
+const (
+	kindName       = "OIDCAdaptor"
+	resultFiltered = "oidcFiltered"
+)
+
+var httpCli = &http.Client{
+	Timeout: 10 * time.Second,
+	CheckRedirect: func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	},
+}
+
+var kind = &filters.Kind{
+	Name:        kindName,
+	Description: "OIDCAdaptor implement OpenID Connect authorization code flow spec",
+	Results:     []string{resultFiltered},
+	DefaultSpec: func() filters.Spec {
+		return &Spec{
+			setAccessTokenHeader: true,
+			setIDTokenHeader:     true,
+			setUserInfoHeader:    true,
+		}
+	},
+	CreateInstance: func(spec filters.Spec) filters.Filter {
+		return &Filter{spec: spec.(*Spec)}
+	},
+}
+
+type store interface {
+	put(key, value string, timeout time.Duration) error
+	get(key string) string
+}
+
+type Filter struct {
+	spec *Spec
+	store
+}
+
+type Spec struct {
+	filters.BaseSpec `yaml:",inline"`
+
+	CookieName string `json:"cookieName"`
+
+	ClientId     string `json:"clientId" jsonschema:"required"`
+	ClientSecret string `json:"clientSecret" jsonschema:"required"`
+
+	//Discovery is oidc discovery standard for example: https://accounts.google.com/.well-known/openid-configuration
+	Discovery string `json:"discovery"`
+
+	//If Discovery not configured, following should be configured for OAuth2 to get userinfo
+	AuthorizationEndpoint string `json:"authorizationEndpoint"`
+	TokenEndpoint         string `json:"tokenEndpoint"`
+	UserinfoEndpoint      string `json:"userinfoEndpoint"`
+
+	RedirectURI string `json:"redirectURI" jsonschema:"required"`
+
+	//Following are user custom defined
+	jwksRefreshInterval  string
+	setAccessTokenHeader bool
+	setIDTokenHeader     bool
+	setUserInfoHeader    bool
+
+	redirectPath string
+	oidcConfig   *oidcConfig
+	jwks         *keyfunc.JWKS
```

Comment:
`Spec` is mainly for static configurations, these fields can be defined in the Filter.